src/daft-sql/src/modules/utf8.rs

```diff
@@ -1,313 +1,509 @@
 };
 
 use super::SQLModule;
 use crate::{
-    ensure,
     error::{PlannerError, SQLPlannerResult},
     functions::{SQLFunction, SQLFunctionArguments},
-    invalid_operation_err, unsupported_sql_err,
+    invalid_operation_err,
 };
 
+macro_rules! utf8_function_one_argument {
+    ($name:ident, $sql_name:expr, $func:expr, $doc:expr, $arg_name:expr) => {
+        pub struct $name;
+
+        impl SQLFunction for $name {
+            fn to_expr(
+                &self,
+                inputs: &[sqlparser::ast::FunctionArg],
+                planner: &crate::planner::SQLPlanner,
+            ) -> SQLPlannerResult<ExprRef> {
+                match inputs {
+                    [input] => {
+                        let input = planner.plan_function_arg(input)?;
+                        Ok($func(input))
+                    }
+                    _ => invalid_operation_err!(concat!(
+                        "invalid arguments for ",
+                        $sql_name,
+                        ". Expected ",
+                        $sql_name,
+                        "(",
+                        $arg_name,
+                        ")"
+                    )),
+                }
+            }
+
+            fn docstrings(&self, _alias: &str) -> String {
+                $doc.to_string()
+            }
+
+            fn arg_names(&self) -> &'static [&'static str] {
+                &[$arg_name]
+            }
+        }
+    };
+}
+
+macro_rules! utf8_function_two_arguments {
+    ($name:ident, $sql_name:expr, $func:expr, $doc:expr, $arg_name_1:expr, $arg_name_2:expr) => {
+        pub struct $name;
+
+        impl SQLFunction for $name {
+            fn to_expr(
+                &self,
+                inputs: &[sqlparser::ast::FunctionArg],
+                planner: &crate::planner::SQLPlanner,
+            ) -> SQLPlannerResult<ExprRef> {
+                match inputs {
+                    [input1, input2] => {
+                        let input1 = planner.plan_function_arg(input1)?;
+                        let input2 = planner.plan_function_arg(input2)?;
+                        Ok($func(input1, input2))
+                    }
+                    _ => invalid_operation_err!(concat!(
+                        "invalid arguments for ",
+                        $sql_name,
+                        ". Expected ",
+                        $sql_name,
+                        "(",
+                        $arg_name_1,
+                        ", ",
+                        $arg_name_2,
+                        ")"
+                    )),
+                }
+            }
+
+            fn docstrings(&self, _alias: &str) -> String {
+                $doc.to_string()
+            }
+
+            fn arg_names(&self) -> &'static [&'static str] {
+                &[$arg_name_1, $arg_name_2]
+            }
+        }
+    };
+}
+
+macro_rules! utf8_function_three_arguments {
```

Comment:
Three macros are ugly, I haven't been able to find better solutions. I would appreciate any assistance or suggestions.

Reply:
I think if we created a few helper functions: 

```rs
fn utf8_unary(
    func: impl Fn(ExprRef) -> ExprRef,
    sql_name: &str,
    arg_name: &str,
    inputs: &[sqlparser::ast::FunctionArg],
    planner: &crate::planner::SQLPlanner,
) -> SQLPlannerResult<ExprRef> {
    match inputs {
        [input] => {
            let input = planner.plan_function_arg(input)?;
            Ok(func(input))
        }
        _ => invalid_operation_err!(
            "invalid arguments for {sql_name}. Expected {sql_name}({arg_name})",
        ),
    }
}

fn utf8_binary(...) 
fn utf3_ternary(...)
```

that'd allow us to simplify the macros into a single one pretty easily. It also pulls the bulk of the code out of macro which will cut down on code generation during compiling. 

with the helper functions, you can create a single macro that matches on all 3 patterns 
```rs
macro_rules! utf8_function {
    ($name:ident, $sql_name:expr, $func:expr, $doc:expr, $arg_name:expr) => {
        // ...
        impl SQLFunction for $name {
            fn to_expr(..) -> SQLPlannerResult<ExprRef> {
                utf8_unary($func, $sql_name, $arg_name, inputs, planner)
            }
            // ...
        }
    };
    ($name:ident, $sql_name:expr, $func:expr, $doc:expr, $arg_name_1:expr, $arg_name_2:expr) => {
        // ...
        impl SQLFunction for $name {
            fn to_expr(..) -> SQLPlannerResult<ExprRef> {
                utf8_binary($func, $sql_name, $arg_name_1, $arg_name_2, inputs, planner)
            }
            // ...
        }
    };
    ($name:ident, $sql_name:expr, $func:expr, $doc:expr, $arg_name_1:expr, $arg_name_2:expr, $arg_name_3:expr) => {
        // ...
        impl SQLFunction for $name {
            fn to_expr(..) -> SQLPlannerResult<ExprRef> {
                utf8_ternary($func, $sql_name, $arg_name_1, $arg_name_2, $arg_name_3, inputs, planner)
            }
            // ...
        }
    };
}
```

Reply:
Cool, I have updated it based on your suggestions.